### PR TITLE
Add conformance test for replace on change component

### PR DESCRIPTION
**Pull request by spider-yamet**
**What**
Fix the provider-replace-on-changes-component conformance test so components with replaceOnChanges have ReplaceOnChanges set in the deployment snapshot.
(Resolve: https://github.com/pulumi/pulumi/issues/21625)

**Why**
The test expects ReplaceOnChanges to be present in the snapshot for components that use the replaceOnChanges option, but the snapshot currently has ReplaceOnChanges as nil. This is needed so remote components with replaceOnChanges correctly drive replace behavior in subsequent updates.

**How**
Ensure the ConformanceComponentProvider forwards ReplaceOnChanges from req.Options.ReplaceOnChanges in its RegisterResource call, and that the engine passes it through to the goal and into the persisted resource state.

### CODE REVIEW

**Comment by spider-yamet** [comment]
@iwahbe @Frassle Conformance test for ignore changes are already passing the test, I opened a test for replace on change component.
Please add impact/no-changelog-required label and check running pipeline status.

Regards

**Comment by spider-yamet** [comment]
@iwahbe @Frassle Could you please run the pipelines? :)

**Comment by Frassle** [comment]
/run-acceptance-tests
Please view the results of the acceptance tests [Here](https://github.com/pulumi/pulumi/actions/runs/22159119208)

**Comment by spider-yamet** [comment]
could you please run the pipelines, @Frassle @iwahbe ?

**Comment by Frassle** [comment]
/run-acceptance-tests
Please view the results of the acceptance tests [Here](https://github.com/pulumi/pulumi/actions/runs/22177052638)

**Comment by spider-yamet** [comment]
could you run again? :) I updated by assertion expect. @Frassle @iwahbe ?

**Comment by Frassle** [comment]
hmm I'm not sure this is well specified actually. We don't currently have a normal conformance test for what replaceOnChanges does for a component resource. Probably want that before worrying about if Construct handles it correctly.
We'll have to think about that.

**Comment by spider-yamet** [comment]
@Frassle Please check, for component resources, the step generator now computes a real diff (old vs new inputs, with ChangedKeys and DetailedDiff) instead of a generic “diffs exist” result. So when a component has replaceOnChanges and a listed property changes, applyReplaceOnChanges runs and the engine correctly produces a replace step.

**Comment by spider-yamet** [comment]
could you run the pipelines once you check?

**Comment by pgavlin** [comment]
@spider-yamet we appreciate your contributions, but it is not productive to repeatedly ping us for reviews. Humans maintain this software; please respect their time accordingly.

File: pkg/testing/pulumi-test-language/tests/provider_replace_on_changes_component.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+	"github.com/pulumi/pulumi/pkg/v3/resource/deploy"
+	"github.com/pulumi/pulumi/pkg/v3/testing/pulumi-test-language/providers"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource/plugin"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func init() {
+	LanguageTests["provider-replace-on-changes-component"] = LanguageTest{
```
Comment (Frassle) [inline pkg/testing/pulumi-test-language/tests/provider_replace_on_changes_component.go]: Before testing this against a language provider we should have a plain l2 test using the "component" provider. Probably best done by going through each of the "l2-resource-option" tests and adding a component resource to each of these tests, that or we duplicate all those tests so we have like "l2-resource-option-replace-on-changes" and "l2-component-option-replace-on-changes". That's probably neater in terms of more fine grained testing, but is a lot more test files.
cc @pulumi/platform on thoughts for that as well.

File: pkg/resource/deploy/step_generator.go [hunk 2]
```diff
@@ -2675,13 +2675,25 @@ func (sg *stepGenerator) diff(
...
 	}
 
-	// If there is no provider for this resource (which should only happen for component resources), simply return a
-	// "diffs exist" result.
+	// If there is no provider for this resource (which should only happen for component resources), compute the
+	// actual diff between old and new inputs so that applyReplaceOnChanges (in continueStepsFromDiff) can correctly
+	// convert changed properties that are in ReplaceOnChanges into a replace step.
 	if prov == nil {
-		if oldInputs.DeepEquals(newInputs) {
+		effectiveNewInputs := processIgnoreChanges(sg.deployment.Diag(), urn, newInputs, oldInputs, goal.IgnoreChanges)
```
Comment (Frassle) [inline pkg/resource/deploy/step_generator.go]: I'm not sure IgnoreChanges _should_ apply to component resources. It can't affect their behaviour, so probably should change the state data saved.

File: pkg/resource/deploy/step_generator.go [hunk 3]
```diff
@@ -2675,13 +2675,25 @@ func (sg *stepGenerator) diff(
...
+		if oldInputs.DeepEquals(effectiveNewInputs) {
```
Comment (Frassle) [inline pkg/resource/deploy/step_generator.go]: No need to call DeepEquals and then call Diff after, just call Diff. Not really saving anything by traversing the tree twice.

File: sdk/go/pulumi-language-go/language_test.go [hunk 4]
```diff
@@ -215,7 +222,7 @@ func TestLanguage(t *testing.T) {
 				SnapshotEdits: []*testingrpc.PrepareLanguageTestsRequest_Replacement{
 					{
 						Path:        "(.+/)?go\\.mod",
-						Pattern:     rootDir + "/",
+						Pattern:     regexp.QuoteMeta(rootDir) + "/",
```
Comment (Frassle) [inline sdk/go/pulumi-language-go/language_test.go:229]: Ah guessing you hit a regex character in your folder path somehow? This fix could be PR'd by itself if you want, nice and small and easy to merge.

File: sdk/go/pulumi-language-go/language_test.go [hunk 5]
```diff
@@ -106,10 +107,9 @@ var expectedFailures = map[string]string{
 	"l1-builtin-can":         "pulumi#18570 Support can in Go program generation",
 
 	// pulumi/pulumi#18345
-	"l1-keyword-overlap":                  "outputs are not cast correctly from pcl to their pulumi types",                                           //nolint:lll
-	"l2-plain":                            "cannot use &plain.DataArgs{…} (value of type *plain.DataArgs) as plain.DataArgs value in struct literal", //nolint:lll
-	"l2-map-keys":                         "cannot use &plain.DataArgs{…} (value of type *plain.DataArgs) as plain.DataArgs value in struct literal", //nolint:lll
-	"l2-discriminated-union":              "pulumi#21829: does not compile",
+	"l1-keyword-overlap":                  "outputs are not cast correctly from pcl to their pulumi types",                                                 //nolint:lll
+	"l2-plain":                            "cannot use &plain.DataArgs{…} (value of type *plain.DataArgs) as plain.DataArgs value in struct literal",       //nolint:lll
+	"l2-map-keys":                         "cannot use &plain.DataArgs{…} (value of type *plain.DataArgs) as plain.DataArgs value in struct literal",       //nolint:lll
```
Comment (Frassle) [inline sdk/go/pulumi-language-go/language_test.go:115]: Need a rebase on master?

File: pkg/testing/pulumi-test-language/tests/provider_replace_on_changes_component.go (hunk 1 above)
Comment (spider-yamet) [reply pkg/testing/pulumi-test-language/tests/provider_replace_on_changes_component.go]: Let me keep a separate component-focused test case (provider-replace-on-changes-component), now colocated in the L2 test file, instead of mixing it into the base l2-resource-option-replace-on-changes assertions.

File: pkg/resource/deploy/step_generator.go (hunk 3 above)
Comment (spider-yamet) [reply pkg/resource/deploy/step_generator.go]: Updated it, revert call DeepEquals

File: pkg/testing/pulumi-test-language/tests/l2_resource_option_replace_on_changes.go [hunk 6]
```diff
@@ -143,4 +145,64 @@ func init() {
 			},
 		},
 	}
+
+	LanguageTests["provider-replace-on-changes-component"] = LanguageTest{
```
Comment (iwahbe) [inline pkg/testing/pulumi-test-language/tests/l2_resource_option_replace_on_changes.go]: We want one file per test, and one test per file. This should go in a file named `provider_replace_on_changes_component`.

**Comment by spider-yamet** [comment]
Now could you please check, @iwahbe @Frassle ?
Update the details, please let me know your opinion.

**Comment by Frassle** [comment]
So there's quite a few merge diffs against master now that need resolving.
But the bigger issue is this needs to be an l2- test now, not a provider- test. Once we've enforced correct behaviour via an l2- test we can write the provider- test. But I need to find time to check with the team on how we want to do those l2- tests, its not clear if its better to cram them into the current option tests, or if we should have separate ones (which would then more easily map to provider- tests). So I'd park this PR for now till we've had time internally to work out what direction we want, that'll probably be sometime next week.

**Comment by spider-yamet** [comment]
@Frassle could you please let me know the solution on l2 test for this resource?
I see lots of PR remaining opening status, hope we can complete one by one.

Regards

**Comment by Frassle** [comment]
So we talked internally and figured it was best to just add component resources to the current l2-resource-option tests. So there's already pkg/testing/pulumi-test-language/tests/testdata/l2-resource-option-ignore-changes/main.pp but it just tests a custom resource, we'll want to add a local component, and remote component resource to that test. We've decided to upgrade local components from l3 to l2 given every language now supports them. We'll want to do similar fixes to all the other option tests as well.

**Comment by spider-yamet** [comment]
May I proceed with just add component resources to the current l2-resource-option tests?

**Comment by spider-yamet** [comment]
@Frassle I added component tests to the current existing l2 test, and tested on my local.
Please check on your side.

Regards

File: pkg/testing/pulumi-test-language/tests/testdata/l2-resource-option-replace-on-changes/0/main.pp [hunk 7]
```diff
@@ -62,3 +62,16 @@
         replaceOnChanges = [value, replaceProp]
     }
 }
+
+// Remote component with replaceOnChanges
+resource "remoteWithReplace" "conformance-component:index:Simple" {
```
Comment (Frassle) [inline pkg/testing/pulumi-test-language/tests/testdata/l2-resource-option-replace-on-changes/0/main.pp]: This shouldn't be using conformance-component, it should be using a component from one of the other builtin providers.

Comment (spider-yamet) [reply pkg/testing/pulumi-test-language/tests/testdata/l2-resource-option-replace-on-changes/0/main.pp]: @Frassle Could you please check now?
